### PR TITLE
adapter,compute: improved dataflow as-of bootstrapping

### LINKED ISSUE #28440: Bootstrap as-of selection may cause dataflows to not be immediately hydrateable
When envd starts up and installs dataflows for all indexes and MVs defined in the catalog, it needs to perform as-of selection, to determine correct and useful initial as-of frontiers for these dataflows. Currently this process takes into account correctness requirements, configured compaction windows, and requirements imposed by compute reconciliation. It also tries to produce as-ofs that are less than the input write frontiers, so that dataflows can hydrate immediately, but it doesn't always succeed.

Here is an example scenario where the resulting as-ofs don't allow one of the dataflows to hydrate immediately:

* dependency tree:
```
Src1 --> Idx1 ---> Idx12
Src2 --> Idx2 -/
```
* persisted (read, write) frontiers:
  * `Src1`: (10, 100)
  * `Src2`: (10, 11)

In its current form, as-of selection will derive the following as-ofs for the three indexes:
* `Idx1`: 99 (input write frontier - compaction window)
* `Idx2`: 10 (input write frontier - compaction window)
* `Idx12`: 99 (least valid read of the inputs)

`Idx12` cannot hydrate immediately but has to wait until `Src2` and `Idx2` have made progress up to time 99.

This is usually fine as input collections tend to catch up quickly after startup. However, for read-only environments running as part of 0dt deployments, this might not always be true. For example, storage collections created by the read-only environment during builtin object migration will not advance their frontiers as long as the environment remains in read-only mode, thus possibly preventing their dependent dataflows from hydrating. We'd like to avoid that since 0dt upgrades can't make progress until all dataflows in the read-only environment have hydrated (or some timeout is reached).

[Relevant Slack thread.](https://materializeinc.slack.com/archives/C071VTZS1M5/p1721405792793109)

### PR DESCRIPTION
**Pull request by teskje**
This PR is a refactor of the as-of bootstrapping process. Currently that logic resides in the coordinator and is split between two methods, `collect_dataflow_storage_constraints` and `bootstrap_dataflow_as_of`, and has some issues:

* It's hard to follow because as-of constraints apply in both directions (downstream and upstream) but `bootstrap_dataflow_as_of` only receives dataflows one at a time, in dependency order. `collect_dataflow_storage_constraints` exists to work around this limitation, supplying information about downstream constraints.
* It is not always able to select as-ofs at which the respective dataflows can hydrate immediately (#28440). This follows directly from the limited insight in constraints imposed by downstream objects and is an issue for 0dt.
* It's implemented in adapter code, even though the cluster team is the de-facto owner of the logic.

This PR tries to address all of these concerns by:

* Moving the as-of selection process into a cluster crate, specifically `mz-compute-client`.
* Supplying the as-of selection process with the full set of dependent dataflows, rather than one at a time.
* Changing the as-of selection process to be more explicit about (a) the current lower and upper bounds for the as-of of each compute collection and (b) the constraints applied to these bounds.

The hope is that the result is more readable and easier to change than what we had before.

### Motivation

  * This PR fixes a recognized bug.

Fixes #28440 

### Tips for reviewer

I had considered making bootstrap as-of selection transparent to adapter by moving it into the controller and running it on `initialization_complete`. That approach ran into issues because adapter wants to take read holds before calling `initialization_complete` and we can't provide them when we don't know the as-ofs yet. This is fixable (e.g. by changing adapter to take read holds later) but it's not clear if the added complexity is worth it.

The new as-of selection process should be mostly equivalent in the outcome to the old one, even though its implementation looks considerably different. The only functional change should be that we do a better job of propagating constraints between dependent dataflows now. In particular, the "warmup" constraint is now effective in ensuring that all collections can immediately hydrate (provided the storage frontiers allow that), whereas it couldn't always ensure that before.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/compute-client/src/as_of_selection.rs [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+    }
+
+    /// Propagate as-of bounds through the dependency graph, in downstream direction.
+    ///
+    /// Calling this method restores `AsOfBounds` invariant (2).
+    fn propagate_bounds_downstream(&self, bound_type: BoundType, constraint_type: ConstraintType) {
+        // We don't want to rely on a correspondence between `GlobalId` order and dependency order,
+        // so we use a fixpoint loop here.
+        fixpoint(|changed| {
+            self.propagate_bounds_downstream_inner(bound_type, constraint_type, changed);
```
Comment (ggevay) [inline src/compute-client/src/as_of_selection.rs:575]: In the worst case, the run time of these fixpoint loops will be kinda quadratic with the number of dataflows. More specifically:
- If there is a long chain of dataflows on a cluster, where each dataflow depends on the previous one, then the fixpoint loop will do as many steps as the length of this dataflow.
- Each step of the fixpoint loop will so as many steps as the number of dataflows across all clusters.

I guess this is ok for now, because this will run in an acceptable time even with 1000 dataflows, which is much more than any of our current users. But in the long run, we might want to improve this in some way. For example, `changed` could be made not just a bool, but it could be a set of dataflows whose bounds have changed in the last step, and then the outer loop of `propagate_bounds_downstream_inner` could go through not all collections, but just those that have an input that changed in the last step.

**Review by ggevay — COMMENTED** [review]
I like the approach! I'm still going through it, but have some preliminary comments.

File: src/compute-client/src/as_of_selection.rs [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+
+            // Propagating `upper` bounds downstream might break `AsOfBounds` invariant (2), so we
+            // need to restore it.
+            if bound_type == BoundType::Upper {
+                self.propagate_bounds_upstream_inner(
+                    BoundType::Upper,
+                    ConstraintType::Hard,
```
[the remark this answers is not in the document]
Comment (teskje) [reply src/compute-client/src/as_of_selection.rs:582]: This is supposed to restore one of the invariants. If it fails, that means one of the invariants is violated and we will probably end up with dataflows that can't hydrate or skip times in their output, which would be a correctness bug.

File: src/compute-client/src/as_of_selection.rs (hunk 1 above)
Comment (teskje) [reply src/compute-client/src/as_of_selection.rs:575]: While we don't assume `GlobalId`s to be in dependency order (I think currently they always are) we can assume them to be mostly in dependency order. So I would expect each fixpoint loop only to be stepped twice (once for the changes, once to notice there are no more changes). An exception is when propagating bounds breaks one of the invariants and we need to propagate in the other direction to fix that. In this case we can iterator multiple times, depending on the shape of dataflow dependencies.

So yes, strictly speaking the fixpoint loops are quadratic (or maybe even worse with the nested fixpoint loops?). I'd hold off with optimizing this until we know it is a performance issue, since it will make the code more complicated. But seems like a good idea to ensure that coordinator bootstrapping measures the time of as-of bootstrapping (it doesn't in the current version).

File: src/compute-client/src/as_of_selection.rs [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+        }
+    }
+
+    /// Apply as-of constraints imposed by the frontiers of upstream storage collections.
+    ///
+    /// A collection's as-of _must_ be >= the read frontier of each of its (transitive) storage
+    /// inputs.
+    ///
+    /// Failing to apply this constraint to a collection is an error. The affected dataflow will
+    /// not be able to hydrate successfully.
```
[the remark this answers is not in the document]
Comment (teskje) [reply src/compute-client/src/as_of_selection.rs:393]: In the scope of compute we just return an error. Adapter then just unwraps it, so yes we will panic. Didn't want to spell it out that explicitly because we might decide to handle this differently in the future. For example, you could imaging marking dataflows as "failed" instead.

File: src/compute-client/src/as_of_selection.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ggevay) [reply src/compute-client/src/as_of_selection.rs:582]: Ah yes, ok. (When I wrote this comment, I mistakenly thought that `AsOfBounds` tracks soft and hard bounds separately, and then propagating soft constrains would have affected only soft bounds, so invariants would have had to be restored only for the soft bounds. But actually, this is not how this works; the soft-hard thing is only for constraints.)

[the remark this answers is not in the document]
Comment (teskje) [reply src/compute-client/src/as_of_selection.rs:582]: Good question! I've convinced myself that this can't happen, but let me try to convince you too.

Let's assume we have just two collections, `A --> B`

* When we start, both invariants are upheld, i.e. `A.lower <= B.lower` and `A.upper <= B.upper`.
* Case 1: soft constraint increases `A.lower`
  * We know that `A.lower` can only be increased up to `A.upper`, so `A.lower' <= A.upper <= B.upper`.
  * Since `A.lower' <= B.upper`, we can propagate the new lower bound to `B`, restoring the invariant.
* Case 2: soft constraint decreases `A.upper`
  * `A.upper' < A.upper <= B.upper`, so the invariant is still upheld without any propagation.
* Case 3: soft constraint increases `B.lower`
  * `A.lower <= B.lower < B.lower'`, so the invariant is still upheld without any propagation.
* Case 4: soft constraint decreases `B.upper`
  * We know that `B.upper` can only be decreased up to `B.lower`, so `A.lower <= B.lower <= B.upper'`.
  * Since `A.lower <= B.upper'`, we can propagate the new upper bound to `A`, restoring the invariant.

Unless I missed something, this means that applying a soft constraint is always safe, as long as one only applies a single constraint at a time and then propagates. We are not doing that, we are always applying one constraint to all collections and then propagate. I think this is still fine as long as a batch of constraint application targets the same bound type (lower or upper).

* Case 5: soft constraint increases both `A.lower` and `B.lower`
  * Case 5.a: `A.lower' <= B.lower'`
    *  The invariant is still upheld without propagation.
  * Case 5.b: `A.lower' > B.lower'`
    * Same as case 1.
* Case 6: soft constraint decreases both `A.upper` and `B.upper`
  * Case 6.a: `A.upper' <= B.upper'`
    *  The invariant is still upheld without propagation.
  * Case 6.b: `A.upper' > B.upper'`
    * Same as case 4.

[the remark this answers is not in the document]
Comment (ggevay) [reply src/compute-client/src/as_of_selection.rs:582]: Ok, I'm convinced! (But the first sentence of the proof seems a bit misleading: I think we don't really want to assume that we have just two collections. Instead, we are picking an arbitrary dependency edge `A --> B`, and proving that after a constraint changes the bounds on `A` or `B`, propagation will always succeed.)

Also, the proof seems to apply not just to these extra `ConstraintType::Hard` calls in `propagate_bounds_downstream` and `propagate_bounds_upstream`, which were the subject of my original comment, but also apply to all `apply_constraint` calls that happen in `propagate_bounds_downstream_inner` and `propagate_bounds_upstream_inner`, right? In other words, no constraint application can ever fail inside any of the `propagate_*` functions. If this is true, then could we remove the `constraint_type` argument of the `propagate_*` functions and make all `apply_constraint` calls in `propagate_*` functions with a `Hard` constraint? 

(Or maybe even create an even harder constraint type, whose failing application would signal an internal error specifically in this module. This would be a more specific error than a normal hard constraint application fail, because a normal hard constraint application fail might also result from, say, corruption in the persisted collection state, if I'm not mistaken.)

[the remark this answers is not in the document]
Comment (teskje) [reply src/compute-client/src/as_of_selection.rs:582]: > (But the first sentence of the proof seems a bit misleading: I think we don't really want to assume that we have just two collections. Instead, we are picking an arbitrary dependency edge `A --> B`, and proving that after a constraint changes the bounds on `A` or `B`, propagation will always succeed.)

Yeah, sorry. I meant to say that this holds for arbitrary dependent collections, and thereby also holds for arbitrary dependency graphs.

> If this is true, then could we remove the `constraint_type` argument of the `propagate_*` functions and make all `apply_constraint` calls in `propagate_*` functions with a `Hard` constraint?

Oh, nice catch! Yeah, I think we could. I'll just make them `Hard` constraints instead of introducing a new constraint type, since the `reason` should already give us specific errors.

[the remark this answers is not in the document]
Comment (ggevay) [reply src/compute-client/src/as_of_selection.rs:582]: And could you add the proof as a code comment?

**Review by antiguru — APPROVED** [review]
I think this looks good! At least, the structure is much easier to understand. I didn't check each of the constraints in detail, but it looks like it's doing what it promises. I left some minor comments in-line.

We should spend some effort on validating it behaves as expected. Certainly a nightly run, but could we let both implementations coexist and validate they produce equal, or explain any different output? Please don't take this as a hard requirement because it might be hard/impossible to have both in parallel!

Another thought: It would be great if we could test this code independently of running all of Mz. Might be hard, too, so not blocking on it!

Approving because I think it seems fine as-is.

File: src/compute-client/src/as_of_selection.rs [hunk 4]
```diff
@@ -0,0 +1,699 @@
...
+            sealed: false,
+        }
+    }
+}
+
+impl<T: fmt::Debug> fmt::Display for AsOfBounds<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "[{:?}, {:?}]",
```
[the remark this answers is not in the document]
Comment (teskje) [reply src/compute-client/src/as_of_selection.rs:186]: I thought I was using the mathematical notation for an interval, but turns out that `[A, B]` is an interval of real numbers, while `[A..B]` is an interval of integers, so your suggestion is better.

File: src/compute-client/src/as_of_selection.rs [hunk 5]
```diff
@@ -0,0 +1,699 @@
...
+/// Runs `step` in a loop until it stops reporting changes.
+fn fixpoint(mut step: impl FnMut(&mut bool)) {
+    loop {
+        let mut changed = false;
+        step(&mut changed);
+        if !changed {
+            break;
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (teskje) [reply src/compute-client/src/as_of_selection.rs:691]: Ah, I think this existed in the old code because I at one point messed up an inlined fixpoint loop, by having `if changed { break; }` or something similarly stupid. Separating the fixpoint logic from the "application" logic makes it easier to reason about each separately, and also removes some boilerplate. I'd prefer to keep the `fixpoint` function since you don't feel strongly.

File: src/compute-client/src/as_of_selection.rs [hunk 6]
```diff
@@ -0,0 +1,699 @@
...
+
+/// Step back the given frontier, if possible.
```
[the remark this answers is not in the document]
Comment (teskje) [reply src/compute-client/src/as_of_selection.rs]: This method is only used in a single place and does exactly what the single place needs: saturate instead of erroring. I expanded the doc comment to include that.

File: src/compute-client/src/as_of_selection.rs (hunk 1 above)
Comment (teskje) [reply src/compute-client/src/as_of_selection.rs:577]: I added logging for the as-of selection runtime in the coordinator's `bootstrap` code.

File: src/compute-client/src/as_of_selection.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (teskje) [reply src/compute-client/src/as_of_selection.rs]: Made the `propagate_bounds_*` methods always use hard constraints and added the proof to the module docstring.

[the remark this answers is not in the document]
Comment (teskje) [reply src/compute-client/src/as_of_selection.rs]: > If this is true, then could we remove the constraint_type argument of the propagate_* functions and make all apply_constraint calls in propagate_* functions with a Hard constraint?

Turns out I we can't just remove the `constraint_type` from the `propagate_bounds_*` methods. It's true that restoring the invariants should always succeed, but `propagate_bounds_*` lets you also do propagations in the other direction from what restores the invariant. For example, the step that applies warmup constraints applies `Upper` constraints to the collections reading from storage directly, and then propagates the `Upper` bounds _downstream_, i.e. it tries to hold back the as-of of downstream collections to the as-ofs of their inputs. Doing so is not important for the invariants, so it's fine if that fails.

I think we can derive the `constraint_type` from the direction and the `bounds_type` though, based on whether they restore an invariant or not. I.e.:
* downstream, `Lower` => `Hard`
* downstream, `Upper` => `Soft`
* upstream, `Lower` => `Soft`
* upstream, `Upper` => `Hard`

**Review by ggevay — APPROVED** [review]
LGTM!

The unit tests are great!

**Comment by teskje** [comment]
> could we let both implementations coexist and validate they produce equal, or explain any different output?

That's hard because we do expect the output to be different. Specifically, the new code lets dataflows hydrate immediately if possible, the old code sometimes couldn't provide that. We could of course log any deviations between the two but I don't see us scouring the logs after each release to manually validate them.

> It would be great if we could test this code independently of running all of Mz.

Thanks, I was too lazy to add unit tests and this pushed me to write them after all. It's a bit annoying because we need to mock `StorageCollections` (which is possible due to the trait!) and produce dummy `DataflowDescription`s, leading to a bunch of boilerplate code. We could avoid that by having `as_of_selection::run` first extract only the necessary information about dependencies and frontiers from `storage_collections` and `dataflows` into smaller types and then run as-of selection only on those, but I lack the energy to do this right now.

**Comment by def-** [comment]
Please ignore the feature benchmark failure in Nightly, it's a test issue, will fix on main.

**Comment by teskje** [comment]
TFTRs!
